iq/iq.go

```diff
@@ -0,0 +1,221 @@
+}
+
+type application struct {
+	ID string `json:"id"`
+}
+
+type thirdPartyAPIResult struct {
+	StatusURL string `json:"statusUrl"`
+}
+
+// StatusURLResult is a struct to let the consumer know what the response from Nexus IQ Server was
+type StatusURLResult struct {
+	PolicyAction  string `json:"policyAction"`
+	ReportHTMLURL string `json:"reportHtmlUrl"`
+	IsError       bool   `json:"isError"`
+	ErrorMessage  string `json:"errorMessage"`
+}
+
+var localConfig configuration.Configuration
+
+var statusURLResp StatusURLResult
+
+func getPurls(purls []string) (result []packageurl.PackageURL) {
+	for _, v := range purls {
+		purl, _ := packageurl.FromString(v)
+		result = append(result, purl)
+	}
+
+	return
+}
+
+// AuditPackages accepts a slice of purls, public application ID, and configuration, and will submit these to
+// Nexus IQ Server for audit, and return a struct of StatusURLResult
+func AuditPackages(purls []string, applicationID string, config configuration.Configuration) StatusURLResult {
+	localConfig = config
+
+	if localConfig.User == "admin" && localConfig.Token == "admin123" {
+		warnUserOfBadLifeChoices()
+	}
+
+	internalID := getInternalApplicationID(applicationID)
+	newPurls := getPurls(purls)
+	sbom := cyclonedx.ProcessPurlsIntoSBOM(newPurls)
+	statusURL := submitToThirdPartyAPI(sbom, internalID)
+
+	finished := make(chan bool)
+
+	statusURLResp = StatusURLResult{}
+
+	go func() {
+		for {
+			select {
+			case <-finished:
+				return
+			default:
+				pollIQServer(fmt.Sprintf("%s/%s", localConfig.Server, statusURL), finished)
+				time.Sleep(pollInterval)
+			}
+		}
+	}()
+
+	<-finished
+	return statusURLResp
+}
+
+func getInternalApplicationID(applicationID string) (internalID string) {
+	client := &http.Client{}
+
+	req, err := http.NewRequest(
+		"GET",
+		fmt.Sprintf("%s%s%s", localConfig.Server, internalApplicationIDURL, applicationID),
+		nil,
+	)
+
+	req.SetBasicAuth(localConfig.User, localConfig.Token)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusOK {
+		bodyBytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+		var response applicationResponse
+		json.Unmarshal(bodyBytes, &response)
+		return response.Applications[0].ID
+	}
+	return ""
+}
+
+func submitToThirdPartyAPI(sbom string, internalID string) string {
+	client := &http.Client{}
+
+	url := fmt.Sprintf("%s%s", localConfig.Server, fmt.Sprintf("%s%s%s%s", thirdPartyAPILeft, internalID, thirdPartyAPIRight, localConfig.Stage))
+
+	req, err := http.NewRequest(
+		"POST",
+		url,
+		bytes.NewBuffer([]byte(sbom)),
+	)
+
+	req.SetBasicAuth(localConfig.User, localConfig.Token)
+
+	req.Header.Set("Content-Type", "application/xml")
+	req.Header.Set("User-Agent", fmt.Sprintf("nancy-client/%s", buildversion.BuildVersion))
+
+	resp, err := client.Do(req)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
```

Comment:
We do this alot..... we already have a `customerrors.Check` in error.go that basically does this and would clean up the if checks all over in this file. Thoughts on using that??

Reply:
I THINK WE SHOULD USE IT, MOSTLY BECAUSE I GAVE YOU A HARD TIME FOR NOT USING IT :)